### ISSUE TITLE
Cannot update silverblue (grub2-mkconfig error)

### ISSUE THREAD

**Issue by kisccz** [fedora-silverblue/issue-tracker#322]
# Solution

See the article: https://fedoramagazine.org/manual-action-required-to-update-fedora-silverblue-kinoite-and-iot-version-36/

Run:

```
$ sudo find /boot/efi -exec touch '{}' ';'
$ sudo touch /etc/kernel/cmdline 
```

Then update your system as usual.

# Original issue below

This issue tracker is intended only for Silverblue specific issues. We would like to ask you to try to reproduce the issue on a relevant Fedora Workstation release. If you will be able to reproduce there, then please report it in Red Hat Bugzilla or in upstream (preferred for GNOME projects) and not in this issue tracker.

**Describe the bug**
I can't update silverblue. I run rpm-ostree upgrade, then reboot but the update is not implemented. I'm on 36.20220810.0 and want to update to 36.20220812.0 

**To Reproduce**
Please describe the steps needed to reproduce the bug:
1. enter command rpm-ostree upgrade or do it through gnome software
2. reboot
3. grub appears with only two options 36.20220810.0 (cuurent version) and 36.20220809.0 (older version), no 36.20220812.0 (version after update)
4. login, update was not implemented

**Expected behavior**
system to update to newest version 36.20220812.0

**Screenshots**

**OS version:**
```
$ rpm-ostree status -b
State: idle
Warning: failed to finalize previous deployment
         error: Bootloader write config: grub2-mkconfig: Child process exited with code 1
         check `journalctl -b -1 -u ostree-finalize-staged.service`
BootedDeployment:
● fedora:fedora/36/x86_64/silverblue
                  Version: 36.20220810.0 (2022-08-10T00:46:50Z)
                   Commit: 0adbf32421d4bf13c9d691b7ea8019f539f47002988e988078383413bfb73aa5
             GPGSignature: Valid signature by 53DED2CB922D8B8D9E63FD18999F7CBF38AB71F4

```

**Additional context**
I'm only running silverblue, no dual boot

**Comment by kisccz** [comment]
```
$ journalctl -b -1 -u ostree-finalize-staged.service
Aug 12 14:57:40 host systemd[1]: Finished ostree-finalize-staged.service - OSTree Finalize Staged Deploymen>
Aug 12 14:58:07 host systemd[1]: Stopping ostree-finalize-staged.service - OSTree Finalize Staged Deploymen>
Aug 12 14:58:07 host ostree[5979]: Finalizing staged deployment
Aug 12 14:58:10 host ostree[5979]: Copying /etc changes: 25 modified, 3 removed, 88 added
Aug 12 14:58:10 host ostree[5979]: Copying /etc changes: 25 modified, 3 removed, 88 added
Aug 12 14:58:14 host ostree[5979]: error: Bootloader write config: grub2-mkconfig: Child process exited wit>
Aug 12 14:58:14 host systemd[1]: ostree-finalize-staged.service: Control process exited, code=exited, statu>
Aug 12 14:58:14 host systemd[1]: ostree-finalize-staged.service: Failed with result 'exit-code'.
Aug 12 14:58:14 host systemd[1]: Stopped ostree-finalize-staged.service - OSTree Finalize Staged Deployment.
Aug 12 14:58:14 host systemd[1]: ostree-finalize-staged.service: Consumed 4.975s CPU time.
```

**Comment by Zlopez** [comment]
I'm having the same issue with the same ostree image.

**Comment by bobslept** [comment]
I had the same issue. After some searching I found these issues:
https://github.com/coreos/rpm-ostree/issues/3925 
https://github.com/coreos/rpm-ostree/issues/3715
https://bugzilla.redhat.com/show_bug.cgi?id=2096192

After that I used the following to fix the issue:
```
sudo find /boot/efi -exec touch '{}' ';'
rpm-ostree upgrade
sudo ostree admin finalize-staged
systemctl reboot
```

**Comment by mesteryui** [comment]
I have the same issue it's possible fix is restarting the service ostree-finalize-staged.service and rebooting the system

**Comment by Pryka** [comment]
Same issue.

**Comment by TheSamsai** [comment]
I ran into the same issue on two separate systems, one running 36.20220810.0 and the other one running 36.20220812.0. Both systems are quite vanilla and I am not running any special kernels on them.

Running
```
rpm-ostree upgrade
sudo ostree admin finalize-staged
```
got both updated to 36.20220814.0 successfully.

The somewhat scary thing is that I only noticed the failed update when I was looking at an unrelated issue and noticed the failure in rpm-ostree status. Depending on when and how this gets fixed, Silverblue installations may go without system updates without the user noticing. This might deserve its own issue, but I think there should probably be more notification about failure to update.

**Comment by bobslept** [comment]
> The somewhat scary thing is that I only noticed the failed update when I was looking at an unrelated issue and noticed the failure in rpm-ostree status. Depending on when and how this gets fixed, Silverblue installations may go without system updates without the user noticing. This might deserve its own issue, but I think there should probably be more notification about failure to update.

Yeah I noticed it when I got like 3 times the same packages when upgrading. And was like huh didn't I update those packages already.

**Comment by Zlopez** [comment]
> The somewhat scary thing is that I only noticed the failed update when I was looking at an unrelated issue and noticed the failure in rpm-ostree status. Depending on when and how this gets fixed, Silverblue installations may go without system updates without the user noticing. This might deserve its own issue, but I think there should probably be more notification about failure to update.

I agree, I noticed this after 2 days only when trying to layer new package and it wasn't in `rpm-ostree status` after reboot. And then I found out the error with grub config.

If this means that the current ostree image is corrupted and it couldn't be updated this could lead to systems that will not get any new update without user even noticing this happened.

**Comment by Pryka** [comment]
> The somewhat scary thing is that I only noticed the failed update when I was looking at an unrelated issue and noticed the failure in rpm-ostree status. Depending on when and how this gets fixed, Silverblue installations may go without system updates without the user noticing. This might deserve its own issue, but I think there should probably be more notification about failure to update.

Fully agree, I noticed it on my system only because some folks on Fedora Discord was talking about it, I check it just out of curiosity and boom i was affected also.

**Comment by zihadio** [comment]
I am facing the same error. Although I can layer other package in 36.1.5 but not with latest version. 

```
$ rpm-ostree status
State: idle
Warning: failed to finalize previous deployment
         error: Bootloader write config: grub2-mkconfig: Child process exited with code 1
         check `journalctl -b -1 -u ostree-finalize-staged.service`
Deployments:
● fedora:fedora/36/x86_64/kinoite
                  Version: 36.20220810.0 (2022-08-10T00:52:28Z)
               BaseCommit: 3ddbe834b41e05e70ae937b37588252643b095ee1bf5865c9c53761ddec65398
             GPGSignature: Valid signature by 53DED2CB922D8B8D9E63FD18999F7CBF38AB71F4
          LayeredPackages: akmod-nvidia rpmfusion-free-release rpmfusion-nonfree-release xorg-x11-drv-nvidia-cuda

  fedora:fedora/36/x86_64/kinoite
                  Version: 36.1.5 (2022-05-04T18:46:01Z)
               BaseCommit: b1f915998a6d923c1a9180b7a3f68ab779359429b56a7abda27c37cb8064da58
             GPGSignature: Valid signature by 53DED2CB922D8B8D9E63FD18999F7CBF38AB71F4
          LayeredPackages: akmod-nvidia xorg-x11-drv-nvidia-cuda
            LocalPackages: rpmfusion-free-release-36-1.noarch rpmfusion-nonfree-release-36-1.noarch
```
This is the error 
```
$ journalctl -b -1 -u ostree-finalize-staged.service
Aug 14 18:52:25 fedora systemd[1]: Finished ostree-finalize-staged.service - OSTree Finalize Staged Deployment.
Aug 14 18:58:29 fedora systemd[1]: Stopping ostree-finalize-staged.service - OSTree Finalize Staged Deployment...
Aug 14 18:58:29 fedora ostree[2811]: Finalizing staged deployment
Aug 14 18:58:31 fedora ostree[2811]: Copying /etc changes: 9 modified, 1 removed, 83 added
Aug 14 18:58:31 fedora ostree[2811]: Copying /etc changes: 9 modified, 1 removed, 83 added
Aug 14 18:58:41 fedora ostree[2811]: error: Bootloader write config: grub2-mkconfig: Child process exited with code 1
Aug 14 18:58:41 fedora systemd[1]: ostree-finalize-staged.service: Control process exited, code=exited, status=1/FAILURE
Aug 14 18:58:41 fedora systemd[1]: ostree-finalize-staged.service: Failed with result 'exit-code'.
Aug 14 18:58:41 fedora systemd[1]: Stopped ostree-finalize-staged.service - OSTree Finalize Staged Deployment.
Aug 14 18:58:41 fedora systemd[1]: ostree-finalize-staged.service: Consumed 4.302s CPU time.
```

**Comment by benkei-kuruma** [comment]
I posted about this on [reddit](https://www.reddit.com/r/Fedora/comments/woan4n/comment/ikaval9/?utm_source=share&utm_medium=web2x&context=3), but I was able to get around this mysterious issue just now with:

```
rpm-ostree update
sudo grub2-mkconfig
```

And that's it. No idea why it worked, but here I am (for now).

**Comment by rscm** [comment]
same here

will try the workaround

```
❯ rpm-ostree status 
State: idle
Warning: failed to finalize previous deployment
         error: Bootloader write config: grub2-mkconfig: El proceso hijo terminó con el código 1
         check `journalctl -b -1 -u ostree-finalize-staged.service`
Deployments:
● fedora:fedora/36/x86_64/silverblue
                  Version: 36.20220810.0 (2022-08-10T00:46:50Z)
               BaseCommit: 0adbf32421d4bf13c9d691b7ea8019f539f47002988e988078383413bfb73aa5
             GPGSignature: Valid signature by 53DED2CB922D8B8D9E63FD18999F7CBF38AB71F4
          LayeredPackages: acpid akmod-nvidia gnome-tweaks gparted openssl printer-driver-brlaser starship tailscale
                           terminator xboxdrv xorg-x11-drv-nvidia xorg-x11-drv-nvidia-libs
            LocalPackages: rpmfusion-free-release-36-1.noarch rpmfusion-nonfree-release-36-1.noarch

  fedora:fedora/36/x86_64/silverblue
                  Version: 36.20220803.0 (2022-08-03T00:51:08Z)
               BaseCommit: 92428f2a60603f29cb8ef04b9eba39aefb2c93a07e9035d7cc164709d65dbfaa
             GPGSignature: Valid signature by 53DED2CB922D8B8D9E63FD18999F7CBF38AB71F4
          LayeredPackages: acpid akmod-nvidia gnome-tweaks gparted openssl printer-driver-brlaser starship tailscale
                           terminator xboxdrv xorg-x11-drv-nvidia xorg-x11-drv-nvidia-libs
            LocalPackages: rpmfusion-free-release-36-1.noarch rpmfusion-nonfree-release-36-1.noarch
```

**Comment by mikezila** [comment]
I had the same issue with the same symptoms and error messages at the others in this thread. I tried the proposed fixes and was able to get one instance of updating/layer to take, but trying to then do another deployment failed with the same error as before.

**Comment by ugurcansayan** [comment]
> I had the same issue with the same symptoms and error messages at the others in this thread. I tried the proposed fixes and was able to get one instance of updating/layer to take, but trying to then do another deployment failed with the same error as before.

+1 that. They feel like temporary fixes, aka workarounds. Hopefully this issue will get a fix soon. Good luck


If anyone wants to put those workarounds in a single command, you can use the && notation. 

Example-1 from https://github.com/fedora-silverblue/issue-tracker/issues/322#issuecomment-1213295354 
`sudo find /boot/efi -exec touch '{}' ';' && rpm-ostree upgrade && sudo ostree admin finalize-staged`

Example-2 from https://github.com/fedora-silverblue/issue-tracker/issues/322#issuecomment-1214451782
`rpm-ostree update && sudo grub2-mkconfig`

rpm-ostree update is an alias to rpm-ostree upgrade, so you can switch them.

Please note that you can also add `&& systemctl` reboot at the end of those lines to make your computer reboot in the end.

**Comment by xrishox** [comment]
is this bug still present in current deployments or was it just if you grabbed it from like august 10-14?

**Comment by zihadio** [comment]
I see a new update has come.
```
$ rpm-ostree upgrade --check
1 metadata, 0 content objects fetched; 592 B transferred in 5 seconds; 0 bytes content written
Enabled rpm-md repositories: fedora-cisco-openh264 fedora-modular updates-modular updates fedora rpmfusion-free-updates rpmfusion-free rpmfusion-nonfree-updates rpmfusion-nonfree updates-archive
Updating metadata for 'fedora-cisco-openh264'... done
Updating metadata for 'fedora-modular'... done
Updating metadata for 'updates-modular'... done
Updating metadata for 'updates'... done
Updating metadata for 'fedora'... done
Updating metadata for 'rpmfusion-free-updates'... done
Updating metadata for 'rpmfusion-free'... done
Updating metadata for 'rpmfusion-nonfree-updates'... done
Updating metadata for 'rpmfusion-nonfree'... done
Updating metadata for 'updates-archive'... done
Importing rpm-md... done
rpm-md repo 'fedora-cisco-openh264'; generated: 2022-04-07T16:52:38Z solvables: 4
rpm-md repo 'fedora-modular'; generated: 2022-05-04T21:12:01Z solvables: 825
rpm-md repo 'updates-modular'; generated: 2022-08-12T01:56:42Z solvables: 1418
rpm-md repo 'updates'; generated: 2022-08-16T01:24:25Z solvables: 21596
rpm-md repo 'fedora'; generated: 2022-05-04T21:16:11Z solvables: 67992
rpm-md repo 'rpmfusion-free-updates'; generated: 2022-08-15T20:06:25Z solvables: 105
rpm-md repo 'rpmfusion-free'; generated: 2022-05-04T04:48:11Z solvables: 506
rpm-md repo 'rpmfusion-nonfree-updates'; generated: 2022-08-15T20:49:20Z solvables: 66
rpm-md repo 'rpmfusion-nonfree'; generated: 2022-05-04T05:11:55Z solvables: 225
rpm-md repo 'updates-archive'; generated: 2022-08-16T01:54:20Z solvables: 31173
Note: --check and --preview may be unreliable.  See https://github.com/coreos/rpm-ostree/issues/1579
AvailableUpdate:
  SecAdvisories: 1 unknown severity, 1 important
           Diff: 3 upgraded
```
I tried ` rpm-ostree update && sudo grub2-mkconfig`.  Didn't work.
```
$ rpm-ostree status
State: idle
Deployments:
● fedora:fedora/36/x86_64/kinoite
                  Version: 36.20220810.0 (2022-08-10T00:52:28Z)
               BaseCommit: 3ddbe834b41e05e70ae937b37588252643b095ee1bf5865c9c53761ddec65398
             GPGSignature: Valid signature by 53DED2CB922D8B8D9E63FD18999F7CBF38AB71F4
          LayeredPackages: akmod-nvidia rpmfusion-free-release rpmfusion-nonfree-release xorg-x11-drv-nvidia-cuda

  fedora:fedora/36/x86_64/kinoite
                  Version: 36.20220810.0 (2022-08-10T00:52:28Z)
               BaseCommit: 3ddbe834b41e05e70ae937b37588252643b095ee1bf5865c9c53761ddec65398
             GPGSignature: Valid signature by 53DED2CB922D8B8D9E63FD18999F7CBF38AB71F4
          LayeredPackages: akmod-nvidia rpmfusion-free-release rpmfusion-nonfree-release xorg-x11-drv-nvidia-cuda
```

Well.. if we can't update. How will this be fixed? Isn't this a deadlock?

**Comment by Zlopez** [comment]
I still see it with `36.20220815.0 (2022-08-15T00:52:34Z)`.

I'm actually not sure if the update will finish even if the fix is introduced in newer version. If not it potentially means we will have plenty of Silverblue systems, that will just stay on those versions that are broken.

**Comment by basvdlei** [comment]
Looks like grub is trying to write into a read-only etc and failing.

Added the following drop-in:

```
# /etc/systemd/system/ostree-finalize-staged.service.d/override.conf
[Service]
Environment=OSTREE_DEBUG_GRUB2=1
```

After running a `rpm-ostree install` command and rebooting, the unit will show the error that occurred with grub:

```
$ journalctl -b -1 -u ostree-finalize-staged.service 
Aug 17 12:16:57 systemd[1]: Finished ostree-finalize-staged.service - OSTree Finalize Staged Deployment.
Aug 17 12:18:37 systemd[1]: Stopping ostree-finalize-staged.service - OSTree Finalize Staged Deployment...
Aug 17 12:18:37 ostree[14815]: Finalizing staged deployment
Aug 17 12:18:37 ostree[14815]: Copying /etc changes: 27 modified, 1 removed, 90 added
Aug 17 12:18:37 ostree[14815]: Copying /etc changes: 27 modified, 1 removed, 90 added
Aug 17 12:18:38 ostree[14882]: Generating grub configuration file ...
Aug 17 12:18:38 ostree[14901]: /etc/grub.d/10_linux: line 167: /etc/kernel/cmdline: Read-only file system
Aug 17 12:18:38 ostree[14815]: error: Bootloader write config: grub2-mkconfig: Child process exited with code 1
Aug 17 12:18:38 systemd[1]: ostree-finalize-staged.service: Control process exited, code=exited, status=1/FAILURE
Aug 17 12:18:38 systemd[1]: ostree-finalize-staged.service: Failed with result 'exit-code'.
Aug 17 12:18:38 systemd[1]: Stopped ostree-finalize-staged.service - OSTree Finalize Staged Deployment.
Aug 17 12:18:38 systemd[1]: ostree-finalize-staged.service: Consumed 1.511s CPU time.
```

This is the section of `10_linux`:

```
   161	    local cmdline="root=${LINUX_ROOT_DEVICE} ro ${GRUB_CMDLINE_LINUX} ${GRUB_CMDLINE_LINUX_DEFAULT}"
   162	    local -a files=($(get_sorted_bls))
   163	
   164	    if [[ ! -f /etc/kernel/cmdline ]]; then
   165		# anaconda has the correct information to do this during install;
   166		# afterward, grubby will take care of syncing on updates.
   167		echo "$cmdline rhgb quiet" > /etc/kernel/cmdline
   168	    fi
   169	
   170	    for bls in "${files[@]}"; do
```

The `/etc/kernel/cmdline` file does not exist on my machine, maybe creating it is enough?

**Comment by jlebon** [comment]
OK, this history is messy here, but I think essentially there are two separate GRUB issues at hand. The first was [GRUB not handling epoch timestamps on the EFI](https://bugzilla.redhat.com/show_bug.cgi?id=2096192). This was fixed in grub2-2.06-43.fc36, which is already in the latest Silverblue composes.

The second was introduced more recently by https://github.com/rhboot/grub2/commit/0837dcdf17ac0429bafa4dbf063b2a94385c04ca (i.e. https://github.com/rhboot/grub2/pull/105) and it causes issues on libostree systems because we run `grub2-mkconfig` [with `/etc` mounted read-only](https://github.com/ostreedev/ostree/blob/ea2f0c294312240cbf1cc286bd71477d13cca89e/src/boot/ostree-finalize-staged.service#L45-L46). I proposed a fix for that in https://github.com/rhboot/grub2/pull/106. In the meantime, a workaround as suggested above is to manually `touch /etc/kernel/cmdline` before updating.

**Comment by subvert0r** [comment]
> I posted about this on [reddit](https://www.reddit.com/r/Fedora/comments/woan4n/comment/ikaval9/?utm_source=share&utm_medium=web2x&context=3), but I was able to get around this mysterious issue just now with:
> 
> ```
> rpm-ostree update
> sudo grub2-mkconfig
> ```
> 
> And that's it. No idea why it worked, but here I am (for now).

This worked and fixed the issue, but can someone explain WHY this fixes the issue?

**Comment by jlebon** [comment]
It works because `grub2-mkconfig` runs outside of the systemd unit which enforces `/etc` being read-only. See the comment above for more details.

**Comment by afland** [comment]
> It works because `grub2-mkconfig` runs outside of the systemd unit which enforces `/etc` being read-only. See the comment above for more details.

So are all existing silverblue systems frozen on the version where this bug first appeared, unless they use one of those manual fixes? Or can a grub update be pushed separate from ostree's system image stuff?

**Comment by juhp** [comment]
(I don't have any /etc/kernel/cmdline or /etc/efi in my vm.)

> rpm-ostree update
> sudo grub2-mkconfig

Should this be the official workaround?  I guess it is impacting a lot of users.

**Comment by tuomastielinen** [comment]
Is this the right command to update GRUB configuration?
`sudo grub2-mkconfig -o /etc/grub2.cfg`

**Comment by travier** [comment]
I think we should write a Fedora Magazine article about this issue to give it more visibility and clear workaround instructions.
I'll work on it next week.

**Comment by Mo8it** [comment]
@travier I agree with needing an article.

We do also need a test that would prevent this from happening again! I am not an expert, but I can imagine that an issue that prevents you from updating the system is hard (or maybe not possible?) to be fixed without user interaction.

We need an integration test that upgrades a base image, overlays a package on it and see if you then have the latest packages and the overlayed package after the reboot.

**Comment by travier** [comment]
Yes, I think we all agree that we need more testing for Silverblue. We have those kind of tests in Fedora CoreOS for example, but unfortunately due to the way things are setup right now, we can not copy/paste that for Silverblue.

If folks are interested in contributing to make this happen we can start a new issue and discuss our options.

**Comment by user1-github** [comment]
It seems todays snapshot is also affected by this issue. I'm currently on 36.20220820.0 which works well, but after I installed todays snapshot (which only updates one package), it still boots into  36.20220820.0 and the update still appears in Gnome software.

**Comment by rowbawts** [comment]
> It seems todays snapshot is also affected by this issue. I'm currently on 36.20220820.0 which works well, but after I installed todays snapshot (which only updates one package), it still boots into 36.20220820.0 after restart, and the update still appears in Gnome software.
> 
> I really don't want to mess with grub configuration in the command line, so will I simply be able to update to a future snapshot that doesn't have this issue?
> 
> **Edit:** I just typed "rpm-ostree status -b" and I don't see any warnings/errors unlike previous reports.

What you can do as a workaround to not have to mess with the command line is rollback to a known good deployment and wait until a fixed snapshot is released and update directly to that one. At least I believe that should work.

**Comment by Pazuu** [comment]
Would it be necessary to inform users via Twitter (and even cooler Mastodon) that their systems are not being updated, that there is a workaround and that it is being investigated?

**Comment by zihadio** [comment]
@user1-github `sudo grub2-mkconfig` really don't mess with grub. it just rebuilt the config. `rpm-ostree update` should have done that for us. So, I say it's a good workaround for now.

And I don't think this is an issue users should fix manually. It will just put silverblue project in bad light.

**Comment by user1-github** [comment]
@zihadio Yeah, but as I said, if I type ""rpm-ostree status -b", I don't get any grub related warnings/errors unlike what people reported here. Currently, that snapshot from yesterday is still the latest, so if the next update will not apply as well, I think I'll open a new issue. BTW, I was on Fedora forums lately and it seems there's a broader issue with rpm-ostree that affects not just Silverblue, so maybe it's related to that.

**Comment by travier** [comment]
See https://github.com/fedora-silverblue/issue-tracker/issues/322#issuecomment-1221293641 for the draft for a Fedora Magazine article.

I don't think we can fix users systems in any way here. All impacted users will have to update with those commands.

**Comment by KurtSchluss** [comment]
> I don't think we can fix users systems in any way here. All impacted users will have to update with those commands.

After a rollback to a previous version with rpm-ostree, I was able to proceed with the normal update process.
Does anybody else had the same experience?

**Comment by user1-github** [comment]
@KurtSchluss Which snapshots have you dealt with?
A few days ago I did a clean install of Silverblue and updated directly to Snapshot 36.20220820.0 without any issues. But like I said, now there's Snapshot 36.20220822.0 (which only updates one package) and for some reason it didn't apply after restart, so I'm still on 36.20220820.0.

**Comment by KurtSchluss** [comment]
> @KurtSchluss Which snapshots have you dealt with? A few days ago I did a clean install of Silverblue and updated directly to Snapshot 36.20220820.0 without any issues. But like I said, now there's Snapshot 36.20220822.0 (which only updates one package) and for some reason it didn't apply after restart, so I'm still on 36.20220820.0.

I faced the same issue with 36.2022**0820**.0. After rolling back to 36.2022**0725**.0, I was able to directly update to 36.2022**0822**.0.

**Comment by SjoerdB93** [comment]
Just today I did an update using `rpm-ostree upgrade` without any workarounds, and it got successfully build into grub. For reference, I upgraded from 36.20220822.0 to 36.20220824.0. So it **seems** the issue has been fixed over here.

**Comment by user1-github** [comment]
Yeah, so it seems 36.20220822.0 and 36.20220824.0 are unaffected by this issue, but 36.20220820.0 is still affected.
I tried updating from 36.20220820.0 to 36.20220824.0 and it still didn't apply. And btw, unlike before, now "rpm-ostree status" does show grub errors.
So I'm just going to wait for the Fedora Magazine article.

**Comment by apevec** [comment]
Upgrade w/o workaround from 36.20220822.0 did not work for me:

Warning: failed to finalize previous deployment
         error: Bootloader write config: grub2-mkconfig: Child process exited with code 1
         check `journalctl -b -1 -u ostree-finalize-staged.service`
Deployments:
● fedora:fedora/36/x86_64/silverblue
                  Version: 36.20220822.0 (2022-08-22T00:51:43Z)
...

And today upgrade fails on fedora-cisco-openh264 repomd mismatch:

Updating metadata for 'fedora-cisco-openh264'... done
error: Loading sack: Updating rpm-md repo 'fedora-cisco-openh264': cannot update repo 'fedora-cisco-openh264': Cannot download repomd.xml: Cannot download repodata/repomd.xml: All mirrors were tried; Last error: Downloading successful, but checksum doesn't match. Calculated: 402d3374ade9cd9a8a35f738f1e215c1962d6aedc411524d2050d359519474f92c3826779ac518773fb44a88f0275981ec24976c7eb845dc775d146a18de1376(sha512)  Expected: eb59e9e7cbde672c4321d25dcc33f5eeadcf48a6bcf3dc97740e51961ce7230e44d1108a8b93b6dabf194d7ec6ac0ee0476c112cdbc65f7e972f22d5370c89bb(sha512)

**Comment by travier** [comment]
@apevec The second issue is likely unrelated and fixed with a `rpm-ostree cleanup -m` and retry.

**Comment by Pryka** [comment]
> 

Same issue with checksum error, rpm-ostree cleanup -m is not fixing the issue.

**Comment by user1-github** [comment]
@Pryka The last few days I couldn't download openh264 codec from Flathub as well, so maybe that's related?

**Comment by travier** [comment]
The second issue is https://github.com/fedora-silverblue/issue-tracker/issues/340

**Comment by omac777** [comment]
So I guess I haven't been successfully upgrading these past couple of weeks.
I assumed it was.

```
rpm-ostree status -b
State: idle
BootedDeployment:
● fedora:fedora/36/x86_64/silverblue
                  Version: 36.20220810.0 (2022-08-10T00:46:50Z)
               BaseCommit: 0adbf32421d4bf13c9d691b7ea8019f539f47002988e988078383413bfb73aa5
             GPGSignature: Valid signature by 53DED2CB922D8B8D9E63FD18999F7CBF38AB71F4
          LayeredPackages: 'c++-gtk-utils-gtk4-devel' 'gcc-c++' 'ImageMagick-c++-devel' 'qt6-*' adwaita-icon-theme-devel alsa-lib-devel autoconf automake avahi-gobject-devel cairo-gobject-devel chromium clang clapper cmake dbus-devel
                           emacs flatpak-builder gcc gdk-pixbuf2-devel gdk-pixbuf2-xlib-devel gh gimp glade glib-devel gnome-video-effects gnonlin graphene-devel gst123 gstreamer1-devel gstreamer1-plugins-bad-free-devel
                           gstreamer1-plugins-bad-free-extras gstreamer1-plugins-bad-free-fluidsynth gstreamer1-plugins-bad-free-wildmidi gstreamer1-plugins-bad-free-zbar gstreamer1-plugins-base-devel gstreamer1-plugins-base-tools
                           gstreamer1-plugins-good-extras gstreamer1-plugins-good-gtk gstreamer1-rtsp-server-devel gstreamer1-svt-av1 gstreamer1-svt-vp9 gstreamer1-vaapi gstreamermm-devel gtk4-devel gtkmm4.0-devel gtkmm4.0-doc ImageMagick
                           ImageMagick-devel ImageMagick-doc libadwaita-devel libgudev-devel libsoup-devel libsoup3-devel libsqlite3x-devel libtoml-devel libxcb-devel lld meld meson mozilla-openh264 openfortivpn openssl1.1-devel
                           poppler-glib-devel python3-jinja2 python3-pygments python3-toml-adapt python3-typogrify qt6-qt3d-examples qt6-qtbase-examples qt6-qtcharts-examples qt6-qtconnectivity-examples qt6-qtdeclarative-examples
                           qt6-qtmultimedia-examples qt6-qtpositioning-examples qt6-qtquick3d-examples qt6-qtwayland-examples remmina svt-av1 tdlib-devel youtube-dl
davidm@carouselYOW 2022-08-25_16:48:41_EDT : ~
```

**Comment by apevec** [comment]
2nd issue is resolved, today 36.20220822.0 -> 36.20220826.0 still showed error: Bootloader write config
so workaround `ostree admin finalize-staged ` is still needed.

**Comment by zihadio** [comment]
I still have issues.

```
[zihad@fedora zihad]$ rpm-ostree status
State: idle
Warning: failed to finalize previous deployment
         error: Bootloader write config: grub2-mkconfig: Child process exited with code 1
         check `journalctl -b -1 -u ostree-finalize-staged.service`
Deployments:
● fedora:fedora/36/x86_64/kinoite
                  Version: 36.20220824.0 (2022-08-24T19:36:40Z)
                   Commit: e0ecc9c08472a8aee1c5fbd65611770c3bf36646979f2a00e38bdf8f33e0c2f7
             GPGSignature: Valid signature by 53DED2CB922D8B8D9E63FD18999F7CBF38AB71F4

  fedora:fedora/36/x86_64/kinoite
                  Version: 36.1.5 (2022-05-04T18:46:01Z)
                   Commit: b1f915998a6d923c1a9180b7a3f68ab779359429b56a7abda27c37cb8064da58
             GPGSignature: Valid signature by 53DED2CB922D8B8D9E63FD18999F7CBF38AB71F4
[zihad@fedora zihad]$ rpm-ostree upgrade --check
1 metadata, 0 content objects fetched; 592 B transferred in 5 seconds; 0 bytes content written
Note: --check and --preview may be unreliable.  See https://github.com/coreos/rpm-ostree/issues/1579
AvailableUpdate:
        Version: 36.20220826.0 (2022-08-26T11:23:21Z)
         Commit: 3251822c1b1b66bb8d2cbf55a5d21877df3aa33edb73dc168dcc97c5a5c48a6d
   GPGSignature: Valid signature by 53DED2CB922D8B8D9E63FD18999F7CBF38AB71F4
           Diff: 17 upgraded
[zihad@fedora zihad]$ 
```

**Comment by meisme-dev** [comment]
> ostree admin finalize-staged

Second issue is still not resolved for me, I ended up just uninstalling openh264 temporarily.

**Comment by cjao** [comment]
> See [#322 (comment)](https://github.com/fedora-silverblue/issue-tracker/issues/322#issuecomment-1221293641) for the draft for a Fedora Magazine article for wider communication. I'm waiting to confirm that the issue is fully fixed and that the commands are enough to fix it.
> 
> I don't think we can fix users systems in any way here. All impacted users will have to update with those commands.

This bug could be fixed pretty easily in Fedora Workstation using a maintainer script since those run as root. Is there something about Silverblue that technically prevents an automated fix?

**Comment by cjao** [comment]
Does this also affect "clean" ostree-based systems with no layered packages? Would Fedora/Red Hat Core OS systems also be vulnerable?

**Comment by Verhoeckx** [comment]
I had the same issue and the workaround (`sudo ostree admin finalize-staged`) did work for me!

**Comment by meisme-dev** [comment]
I don't get the error anymore, was it fixed?

**Comment by SjoerdB93** [comment]
> I don't get the error anymore, was it fixed?

On my machine it has been fixed since version from 36.20220822.0 or so. I've done three different upgrades since then (the latest being from 36.20220827.0 to from 36.20220828.0) and all went perfect without any errors or workarounds that I applied, just a simple `rpm-ostree upgrade` and nothing else.

**Comment by omac777** [comment]
On another fedora silverblue box with nvidia legacy gpu on it I needed to do this set of steps to succeed:
```
rpm-ostree status
sudo grub2-mkconfig
sync
sudo rpm-ostree upgrade
sync
sudo find /boot/efi -exec touch '{}' ';'
sudo ostree admin finalize-staged
sync
sudo systemctl reboot
```

After the reboot here is the status...SUCCESS...thank you everybody:
```
davidm@io02Silverblue 2022-08-29_10:19:34_EDT : ~
 $ rpm-ostree status
State: idle
Deployments:
● fedora:fedora/36/x86_64/silverblue
                  Version: 36.20220828.0 (2022-08-28T22:08:28Z)
               BaseCommit: a1526635f0c0e9fbd6a2f49f3fea36e90c568224b72e0d02f51dcb34d7cbb917
             GPGSignature: Valid signature by 53DED2CB922D8B8D9E63FD18999F7CBF38AB71F4
          LayeredPackages: 'c++-gtk-utils-gtk4-devel' 'gcc-c++' 'ImageMagick-c++-devel' 'libstdc++-devel' 'qt6-*' adwaita-icon-theme-devel akmod-nvidia-470xx alsa-lib-devel autoconf automake
                           avahi-gobject-devel blender cairo-devel cairo-gobject-devel chromium clang clang-tools-extra clapper cmake dbus-devel emacs gcc gdk-pixbuf2-devel gh gimp
                           glib-devel glib2-devel graphene-devel gst123 gstreamer1-devel gstreamer1-plugins-bad-free-devel gstreamer1-plugins-base-devel gstreamer1-rtsp-server-devel
                           gstreamermm-devel gtk4-devel gtkmm4.0-devel gtkmm4.0-doc ImageMagick ImageMagick-devel ImageMagick-doc libgudev-devel libshaderc-devel libsoup-devel libsoup3-devel
                           libsqlite3x-devel libtoml-devel libva-utils libva-vdpau-driver libxcb-devel lld meld meson nvidia-settings-470xx openfortivpn openssl-devel pango-devel
                           python3-jinja2 python3-pygments python3-toml-adapt python3-typogrify qt6-qt3d-examples qt6-qtbase-examples qt6-qtcharts-examples qt6-qtconnectivity-examples
                           qt6-qtdeclarative-examples qt6-qtmultimedia-examples qt6-qtpositioning-examples qt6-qtquick3d-examples qt6-qtwayland-examples remmina screen svt-av1 vdpauinfo
                           vkmark vulkan-tools xorg-x11-drv-nvidia-470xx-cuda xorg-x11-drv-nvidia-470xx-power xrdp youtube-dl

  fedora:fedora/36/x86_64/silverblue
                  Version: 36.20220818.0 (2022-08-18T01:32:08Z)
               BaseCommit: 7dd29734d2403cb8cd8e86b0af53d9e92de5eff8ad5c22ab9f145131cdaf51b7
             GPGSignature: Valid signature by 53DED2CB922D8B8D9E63FD18999F7CBF38AB71F4
          LayeredPackages: 'c++-gtk-utils-gtk4-devel' 'gcc-c++' 'ImageMagick-c++-devel' 'libstdc++-devel' 'qt6-*' adwaita-icon-theme-devel akmod-nvidia-470xx alsa-lib-devel autoconf automake
                           avahi-gobject-devel blender cairo-devel cairo-gobject-devel chromium clang clang-tools-extra clapper cmake dbus-devel emacs gcc gdk-pixbuf2-devel gh gimp
                           glib-devel glib2-devel graphene-devel gst123 gstreamer1-devel gstreamer1-plugins-bad-free-devel gstreamer1-plugins-base-devel gstreamer1-rtsp-server-devel
                           gstreamermm-devel gtk4-devel gtkmm4.0-devel gtkmm4.0-doc ImageMagick ImageMagick-devel ImageMagick-doc libgudev-devel libshaderc-devel libsoup-devel libsoup3-devel
                           libsqlite3x-devel libtoml-devel libva-utils libva-vdpau-driver libxcb-devel lld meld meson nvidia-settings-470xx openfortivpn openssl-devel pango-devel
                           python3-jinja2 python3-pygments python3-toml-adapt python3-typogrify qt6-qt3d-examples qt6-qtbase-examples qt6-qtcharts-examples qt6-qtconnectivity-examples
                           qt6-qtdeclarative-examples qt6-qtmultimedia-examples qt6-qtpositioning-examples qt6-qtquick3d-examples qt6-qtwayland-examples remmina screen svt-av1 vdpauinfo
                           vkmark vulkan-tools xorg-x11-drv-nvidia-470xx-cuda xorg-x11-drv-nvidia-470xx-power xrdp youtube-dl

  fedora:fedora/35/x86_64/silverblue
                  Version: 35.20220424.0 (2022-04-24T19:18:01Z)
               BaseCommit: f48811fab8df9ed501f2db5d0558d77f12942740d34803f7b76ad02cac5513ef
             GPGSignature: Valid signature by 787EA6AE1147EEE56C40B30CDB4639719867C58F
          LayeredPackages: 'libstdc++-devel' akmod-nvidia-470xx avahi-gobject-devel blender cairo-devel cairo-gobject-devel clang clang-tools-extra emacs gdk-pixbuf2-devel glib-devel
                           glib2-devel graphene-devel gtk4-devel libshaderc-devel libva-utils libva-vdpau-driver mesa-vulkan-devel nvidia-settings-470xx pango-devel vdpauinfo vkmark
                           vulkan-tools x2goserver xorg-x11-drv-nvidia-470xx-cuda xorg-x11-drv-nvidia-470xx-power xrdp
            LocalPackages: rpmfusion-free-release-35-1.noarch rpmfusion-nonfree-release-35-1.noarch
                   Pinned: yes
```

**Comment by jmaibaum** [comment]
I still need `sudo ostree admin finalize-staged` after each `rpm-ostree upgrade`, last time just a few minutes ago.

```
$ rpm-ostree status
State: idle
Deployments:
● fedora:fedora/36/x86_64/silverblue
                  Version: 36.20220828.0 (2022-08-28T22:08:28Z)
               BaseCommit: a1526635f0c0e9fbd6a2f49f3fea36e90c568224b72e0d02f51dcb34d7cbb917
             GPGSignature: Valid signature by 53DED2CB922D8B8D9E63FD18999F7CBF38AB71F4
          LayeredPackages: flatpak-builder gstreamer1-plugin-openh264 openssl simple-scan

  fedora:fedora/36/x86_64/silverblue
                  Version: 36.20220827.0 (2022-08-27T20:13:05Z)
               BaseCommit: 3623cd53d38fa768a777d5d372a101bd5c545e0c1db47c4e62f6eea5effac701
             GPGSignature: Valid signature by 53DED2CB922D8B8D9E63FD18999F7CBF38AB71F4
          LayeredPackages: flatpak-builder gstreamer1-plugin-openh264 openssl simple-scan
```

It seems that there are people who don't need this anymore and I wonder what I have missed?

**Comment by SjoerdB93** [comment]
> It seems that there are people who don't need this anymore and I wonder what I have missed?

One difference I notice is that I ran `sudo grub2-mkconfig` as a workaround instead of  `sudo ostree admin finalize-staged `. I'm not completely sure what the latter does exactly but it may explain some of the difference experiences?

**Comment by zihadio** [comment]
I am still facing issue updating without `sudo grub2-mkconfig`. Is this silverblue issue? Or Grub issue? I see some youtube channel claiming it as a grub issue?

```
❯ rpm-ostree status
State: idle
Deployments:
● fedora:fedora/36/x86_64/kinoite
                  Version: 36.20220828.0 (2022-08-28T22:15:50Z)
               BaseCommit: 2b6a895c510d910549bc5cf3d523a3bf65f1f5666ea6d2ab9b38a0eb1346de4e
             GPGSignature: Valid signature by 53DED2CB922D8B8D9E63FD18999F7CBF38AB71F4
          LayeredPackages: akmod-nvidia distrobox ffmpegthumbs htop neofetch rpmfusion-free-release
                           rpmfusion-nonfree-release xorg-x11-drv-nvidia-cuda zsh

  fedora:fedora/36/x86_64/kinoite
                  Version: 36.20220828.0 (2022-08-28T22:15:50Z)
               BaseCommit: 2b6a895c510d910549bc5cf3d523a3bf65f1f5666ea6d2ab9b38a0eb1346de4e
             GPGSignature: Valid signature by 53DED2CB922D8B8D9E63FD18999F7CBF38AB71F4
          LayeredPackages: akmod-nvidia distrobox ffmpegthumbs neofetch rpmfusion-free-release
                           rpmfusion-nonfree-release xorg-x11-drv-nvidia-cuda zsh

```

**Comment by jwillikers** [comment]
After using the `sudo grub2-mkconfig` / `sudo ostree admin finalize-staged` after upgrading to 36.20220830.0, I'm no longer seeing this issue. I was able to successfully layer a package without any additional steps.

**Comment by jmaibaum** [comment]
> One difference I notice is that I ran `sudo grub2-mkconfig` as a workaround instead of `sudo ostree admin finalize-staged `. I'm not completely sure what the latter does exactly (I always assumed it ran mkconfig by default, but maybe not) but it may explain some of the different experiences?

OK, when this morning's update to `36.20220830.0` again failed, I ran `sudo grub2-mkconfig` as a workaround, instead of `sudo ostree admin finalize-staged`. Will report again if the next update will go through without a workaround or not.

**Comment by crebwyllion** [comment]
I've using `sudo ostree admin finalize-staged` to get my Silverblue upgrades and layerings through, and I was still having the issue through tonight. After I upgraded to `36.20220830.0`, I tried using `sudo grub2-mkconfig` instead. I restarted, my upgrade went through. After that, I tried layering a package, and THAT finally went through without having to use either workaround. We'll see tomorrow if my upgrades go through without workarounds as well.

**Comment by Zlopez** [comment]
I only used `sudo ostree admin finalize-staged` and I didn't needed to use it for the last few updates. Currently I'm on `36.20220828.0`.

**Comment by tmstaedt** [comment]
> # Solution
> Run:
> 
> ```
> $ sudo find /boot/efi -exec touch '{}' ';'
> $ sudo touch /etc/kernel/cmdline 
> ```
> 
> Then update your system as usual.
> 
> # Original issue below

....

Also saw this today. The above two commands fixed it.

**Comment by ugurcansayan** [comment]
Well, I was going with `rpm-ostree upgrade && sudo grub2-mkconfig` until today. Today I only used `rpm-ostree upgrade` and next reboot, update was applied -- as it should be. 

Seeing different workarounds coming to same result, maybe this is just placebo? Or something else to investigate? Idk.

**Comment by travier** [comment]
I've update the first comment with what we think is the correct and sufficient workaround. I'll be testing that to confirm it and then we will publish the Fedora Magazine article with it.

**Comment by travier** [comment]
If folks can install an old affected 36 version (36.20220810.0), try the workaround are report success / failure that would be really useful.

**Comment by Mershl** [comment]
` grub2-2.06-52.fc36` just arrived on F36 and the latest Silverblue 36 deployment. This includes a commit specifying refering this issue:
https://src.fedoraproject.org/rpms/grub2/c/714559fb3d582fb52bfde8589803ee6094857edd?branch=f36

**Comment by bobslept** [comment]
> If folks can install an old affected 36 version (36.20220810.0), try the workaround and report success / failure that would be really useful.

```
BootedDeployment:
● fedora:fedora/36/x86_64/silverblue
                   Version: 36.20220616.0 (2022-06-16T00:47:43Z)
                BaseCommit: 9f21661171d4da930e8ed28cfc7f13b88e457d48efd0c460106c49790495bdf2
              GPGSignature: Valid signature by 53DED2CB922D8B8D9E63FD18999F7CBF38AB71F4
           LayeredPackages: flatpak-module-tools
                    Pinned: yes
```
```
rpm-ostree deploy 36.20220810.0
systemctl reboot
```
```
rpm-ostree status -b
BootedDeployment:
● fedora:fedora/36/x86_64/silverblue
                  Version: 36.20220810.0 (2022-08-10T00:46:50Z)
               BaseCommit: 0adbf32421d4bf13c9d691b7ea8019f539f47002988e988078383413bfb73aa5
             GPGSignature: Valid signature by 53DED2CB922D8B8D9E63FD18999F7CBF38AB71F4
          LayeredPackages: flatpak-module-tools
```
```
rpm-ostree upgrade
systemctl reboot
```
```
rpm-ostree status -b 
Warning: failed to finalize previous deployment
         error: Bootloader write config: grub2-mkconfig: Child process exited with code 1
         check `journalctl -b -1 -u ostree-finalize-staged.service`
BootedDeployment:
● fedora:fedora/36/x86_64/silverblue
                  Version: 36.20220810.0 (2022-08-10T00:46:50Z)
               BaseCommit: 0adbf32421d4bf13c9d691b7ea8019f539f47002988e988078383413bfb73aa5
             GPGSignature: Valid signature by 53DED2CB922D8B8D9E63FD18999F7CBF38AB71F4
          LayeredPackages: flatpak-module-tools
```

```
sudo find /boot/efi -exec touch '{}' ';'
find: ‘/boot/efi’: No such file or directory
```
```
sudo touch /etc/kernel/cmdline 
rpm-ostree upgrade
systemctl reboot
```
```
rpm-ostree status -b
BootedDeployment:
● fedora:fedora/36/x86_64/silverblue
                  Version: 36.20220901.0 (2022-09-01T09:02:25Z)
               BaseCommit: 57b94f7ca31fba480a2a779039b152deb0c24e2b975eb56bc204cac24d4299bb
             GPGSignature: Valid signature by 53DED2CB922D8B8D9E63FD18999F7CBF38AB71F4
          LayeredPackages: flatpak-module-tools
```

Seems to be working on my end.

**Comment by Verhoeckx** [comment]
> I had the same issue and the workaround (`sudo ostree admin finalize-staged`) did work for me!

But it wasn't enough, I today also had to apply `sudo grub2-mkconfig`. The upgrade now seemed to go well.

**Comment by dinhokusanagi** [comment]
For me the solution was this:
rpm-ostree update
sudo grub2-mkconfig

I've updated several times without problems.

**Comment by bobslept** [comment]
I tried another VM, this time with EFI enabled. Fresh install with `rpm-ostree deploy 36.20220810.0` afterwards. Another sucessfull upgrade with current solution:
```
sudo find /boot/efi -exec touch '{}' ';'
sudo touch /etc/kernel/cmdline
```

**Comment by user1-github** [comment]
So what's the difference between "sudo grub2-mkconfig" and the new official workaround?

Yesterday I upgraded to 36.20220831.0 by using the "sudo grub2-mkconfig" workaround.
Today I also updated to 36.20220901.0 which now went smooth without any workaround. And finally I also installed Steam as a layered package, which also went smooth without any workaround.

My rpm-ostree status output:

> State: idle
Deployments:
● fedora:fedora/36/x86_64/silverblue
                  Version: 36.20220901.0 (2022-09-01T09:02:25Z)
               BaseCommit: 57b94f7ca31fba480a2a779039b152deb0c24e2b975eb56bc204cac24d4299bb
             GPGSignature: Valid signature by 53DED2CB922D8B8D9E63FD18999F7CBF38AB71F4
          LayeredPackages: steam

  fedora:fedora/36/x86_64/silverblue
                  Version: 36.20220901.0 (2022-09-01T09:02:25Z)
                   Commit: 57b94f7ca31fba480a2a779039b152deb0c24e2b975eb56bc204cac24d4299bb
             GPGSignature: Valid signature by 53DED2CB922D8B8D9E63FD18999F7CBF38AB71F4

**Comment by jmaibaum** [comment]
> > One difference I notice is that I ran `sudo grub2-mkconfig` as a workaround instead of `sudo ostree admin finalize-staged `. I'm not completely sure what the latter does exactly (I always assumed it ran mkconfig by default, but maybe not) but it may explain some of the different experiences?
> 
> OK, when this morning's update to `36.20220830.0` again failed, I ran `sudo grub2-mkconfig` as a workaround, instead of `sudo ostree admin finalize-staged`. Will report again if the next update will go through without a workaround or not.

Had two `rpm-ostree upgrade` updates, and one `rpm-ostree install ...` package layering going through in the meantime without having to do any workaround, so I assume that either `sudo grub2-mkconfig` from two days ago or today's update of `grub2` fixed it. Didn't try the now officially recommended solution in the ticket description anymore as I didn't need it anymore.

**Comment by bobslept** [comment]
Both of the VM's where I used the current solution updated without any problems today from `36.20220901.0` to `36.20220902.0`

**Comment by travier** [comment]
Thanks @bobslept for your precise instructions in https://github.com/fedora-silverblue/issue-tracker/issues/322#issuecomment-1234157279. I've followed the exact same steps in a fresh VM and verified that this works around the issue.

We'll publish the article soon. Thanks a lot for testing it.

**Comment by travier** [comment]
Article published: https://fedoramagazine.org/manual-action-required-to-update-fedora-silverblue-kinoite-and-iot-version-36/

Closing this one. Thanks for the reports and testing!

**Comment by strugee** [comment]
@travier it would be useful to include the version this was fixed in the article - I somehow never ended up on a bad image, but I want to double-check that I'm rebooting into a fixed image.

**Comment by cjao** [comment]
Are there any measures (e.g. automated tests) to prevent this sort of bug from happening again?

**Comment by travier** [comment]
> Are there any measures (e.g. automated tests) to prevent this sort of bug from happening again?

We're working on it. First step is tracked in https://github.com/fedora-silverblue/issue-tracker/issues/334.

**Comment by travier** [comment]
Someone made a video explaining this issue: https://www.youtube.com/watch?v=BlAgtCYIBQQ

**Comment by rugk** [comment]
Being [absolutely dumb and late to the party I just discovered this today](https://ask.fedoraproject.org/t/fedora-silverblue-rpm-ostree-not-updating-anymore-bootloader-write-config-grub2-mkconfig-error-in-ostree-finalize-staged-service/27343?u=rugk) when I finally checked that update via the CLI.

The issue was GNOME Software (and actually the CLI also not until you run `rpm-ostree status` and read carefully) could not detect this as a failure or showed me _any kind of notification_ that something went wrong.
IMHO this should also be treated as a bug/missing thing as to miss such a thing is really a critical thing and any kind of notification would have caught my attention way earlier.

(did not read the whole issue there, sry if already fixed or so)

**Comment by ptomato** [comment]
I just realized today that I haven't been getting any updates for _months_, with no notification that the updates failed. I'd recommend taking a few actions:
- I see the link to the [Fedora Magazine article](https://fedoramagazine.org/manual-action-required-to-update-fedora-silverblue-kinoite-and-iot-version-36/) is at the very top of this issue, and the issue is pinned in the issue tracker - good thinking! I'd recommend additionally pinning an issue in any other issue trackers where users might look. Personally, I started my search in coreos/rpm-ostree, but there are probably a few more places where people who are somewhat familiar with the technology might start looking.
- I'd recommend pushing some sort of notification to users' gnome-software. I realize that's hard to do if ostree updates are broken, but maybe there is an out-of-band way to do it. You could create a flatpak app whose purpose is to bring this fix to users' attention, add it to Fedora's flatpak repository, and make sure gnome-software features it. (IIRC that is achieved through a key in the flatpak metadata?)
- In future it might be good to have some sort of out-of-band mechanism to push notifications in emergencies like this. e.g. a static URL that usually 404s but serves a notification text if manual action is required.

If it took a tech-savvy user like me, who's somewhat familiar with ostree, months to even realize there was a problem, I'm assuming that non-tech-savvy Silverblue users are just not going to notice and if they do, they won't know where to look.

**Comment by travier** [comment]
I agree, it's really bad. Unfortunately we don't have any "back channel" to push a notification for that right now so it's not possible to do it for this issue.

**Comment by miabbott** [comment]
> If it took a tech-savvy user like me, who's somewhat familiar with ostree, months to even realize there was a problem, I'm assuming that non-tech-savvy Silverblue users are just not going to notice and if they do, they won't know where to look.

I wonder if there is something that can be done in GNOME Software (and the KDE equivalent) where an alert is generated saying something like "you haven't received any OS updates for `<length of time>`" or "it's been `<length of time>` since you applied an OS update"

That way the user could be made aware of a possible problem with the update mechanism or access to the ostree updates.

**Comment by travier** [comment]
That's a good idea. There is logic in Discover to detect EOL'ed versions of the distribution. We could extend that to check if the current release is too old. Something similar could likely be done for GNOME Software.

**Comment by ptomato** [comment]
What do you think of my idea above, of publishing a one-purpose, featured, Flatpak app in the Fedora Flatpak repository, whose purpose is to point users to the fix?

**Comment by miabbott** [comment]
> What do you think of my idea above, of publishing a one-purpose, featured, Flatpak app in the Fedora Flatpak repository, whose purpose is to point users to the fix?

I could see how that would work...though I think I would prefer something more generic in GNOME Software/Discover that may be able to be reused in the future.

One drawback to the singular Flatpak idea is that it would still be an optional install, so some users wouldn't be exposed to the functionality.

Neither idea is perfect, though.

**Comment by miabbott** [comment]
> I wonder if there is something that can be done in GNOME Software (and the KDE equivalent) where an alert is generated saying something like "you haven't received any OS updates for `<length of time>`" or "it's been `<length of time>` since you applied an OS update"

https://gitlab.gnome.org/GNOME/gnome-software/-/issues/1964

**Comment by travier** [comment]
To me the Flatpak option creates a lot of questions:
- How would you name it?
- What would it do?
- How can we show it only to the relevant users?
- If we feature it (can we without updates?), then it will be forced for users that don't have the issue too
- When do we remove it?
- etc.

**Comment by ptomato** [comment]
I agree, it would be a pretty gross and extreme solution, but possibly justified? I see "silently fail all upgrades indefinitely" as a disaster requiring extreme measures...

To answer the specific questions:
- I'd call it something like "Fedora Silverblue Upgrade Rescue Tool" or a more attention-grabbing name if possible. (I'm sure Red Hat has design and marketing professionals that could come up with something better than me!)
- What it would do would be basically to show a window explaining what's going on and giving people the code to copy into their terminal, with links to the relevant issue if they want to verify/understand what they're typing in. Maybe with certain permissions it'd be able to check if their system is suffering from the bug or already fixed; or maybe that wouldn't be possible from inside the Flatpak sandbox at all.
- I doubt it would be able to be shown only to the relevant users, they'd just have to ignore it.
- I assume it'd be able to be featured without an update, because IIRC the metadata for Flatpaks is a separate file served by the Flatpak repository?
- It could be removed anytime down the line, after whatever metrics you have show that usage of 36 is below a certain threshold? The point would be to reach like 3/4 of the people whose upgrades are still blocked; I doubt 100% would be possible.

**Comment by alkazar** [comment]
I don't know if the root cause is the same, but I experienced this exact issue today when trying to upgrade from Fedora Silverblue 38 to 39.

**Comment by francoism90** [comment]
Same here.. it just didn't do anything.

I had to follow https://github.com/fedora-silverblue/issue-tracker/issues/322#issuecomment-1213295354, https://github.com/fedora-silverblue/issue-tracker/issues/322 and https://github.com/fedora-silverblue/issue-tracker/issues/322#issuecomment-1214451782 to get back to a working state.

It seems to loop:
```
nov 11 14:21:48 desktop root[10485]: os-prober: debug: running /usr/libexec/os-probes/mounted/05efi on mounted /dev/nvme0n1p1
nov 11 14:21:48 desktop root[10485]: 05efi: debug: /dev/nvme0n1p1 is a FAT32 partition
nov 11 14:21:48 desktop root[10485]: 05efi: debug: /dev/nvme0n1p1 partition scheme is gpt
nov 11 14:21:48 desktop root[10485]: 05efi: debug: /dev/nvme0n1p1 partition type is c12a7328-f81f-11d2-ba4b-00a0c93ec93b
nov 11 14:21:48 desktop root[10485]: 05efi: debug: running subtest /usr/libexec/os-probes/mounted/efi/05shell
nov 11 14:21:48 desktop root[10485]: 05efi: debug: running subtest /usr/libexec/os-probes/mounted/efi/10elilo
nov 11 14:21:48 desktop root[10485]: 05efi: debug: running subtest /usr/libexec/os-probes/mounted/efi/20microsoft
nov 11 14:21:48 desktop root[10485]: os-prober: debug: running /usr/libexec/os-probes/mounted/10freedos on mounted 
...
```
I could send the whole log, but it would simple do the same thing over-and-over again.
